CSU DSCI 369 Lab 10
Instructor: Emily J. King
Spring 2024

Goals:  Understand geometric interpretation of linear transformations R^2 -> R^2, including invertibility and determinant
Also needed: DSCI369_Sp24_Lab10.pdf

In [1]:
import numpy as np
from numpy.linalg import det
from numpy.linalg import inv

Input the matrices.  

Note that np.cos(np.radians()) and np.sin(np.radians()) are the Numpy commands for cosine and sine with degrees as the unit of inputs.  Also note that np.diag(vector with d entries) yields a dxd diagonal matrix with diagonal entries the entries of the vector.

In [2]:
D1 = np.diag(np.array([0.5, 3]))
D2 = np.diag(np.array([-1, -1]))
D3 = np.diag(np.array([1, -1]))
D4 = np.diag(np.array([0, 1]))
R = np.array([[np.cos(np.radians(30)), -np.sin(np.radians(30))], [np.sin(np.radians(30)), np.cos(np.radians(30))]])
S = np.array([[1, 0], [2, 1]])
A = np.random.rand(2,2)

Analyze D1

Use Worksheet A to determine how multiplying by D1 changes areas.

Use Worksheet B to determine if one needs to flip out of the plane to go from the red F to the blue F.

Compute the determinant of D1.

In [3]:
det(D1)

1.5000000000000002

Why does this make sense? Discuss.
Because the determinant is positive so you don't need to go out of the plane to complete the transformation

Is D1 invertible?  If so, what is its inverse?

In [4]:
inv(D1)

array([[2.        , 0.        ],
       [0.        , 0.33333333]])

Use Worksheets A & B to determine what multiplying by the inverse of D1 does.  Why does this make sense?  Discuss.

Analyze D2

Use Worksheet A to determine how multiplying by D2 changes areas.

Use Worksheet B to determine if one needs to flip out of the plane to go from the red F to the blue F.

Compute the determinant of D2.

In [5]:
det(D2)

1.0

Why does this make sense? Discuss.

Is D2 invertible?  If so, what is its inverse?

In [6]:
inv(D2)

array([[-1., -0.],
       [-0., -1.]])

Use Worksheets A & B to determine what multiplying by the inverse of D1 does.  Why does this make sense?  Discuss.

Analyze D3

Use Worksheet A to determine how multiplying by D3 changes areas.

Use Worksheet B to determine if one needs to flip out of the plane to go from the red F to the blue F.

Compute the determinant of D3.

In [7]:
det(D3)

-1.0

Why does this make sense? Discuss.

Is D3 invertible?  If so, what is its inverse?

In [8]:
inv(D3)

array([[ 1.,  0.],
       [-0., -1.]])

Use Worksheets A & B to determine what multiplying by the inverse of D3 does.  Why does this make sense?  Discuss.

Analyze D4

Use Worksheet A to determine how multiplying by D4 changes areas.

Use Worksheet B to determine if one needs to flip out of the plane to go from the red F to the blue F.

Compute the determinant of D4.

In [9]:
det(D4)

0.0

Why does this make sense? Discuss.

Is D4 invertible?  If so, what is its inverse?

In [23]:
inv(D4)

LinAlgError: Singular matrix

Why does this make sense?  Discuss.

Analyze R

Use Worksheet A to determine how multiplying by R changes areas.

Use Worksheet B to determine if one needs to flip out of the plane to go from the red F to the blue F.

Compute the determinant of R.

In [11]:
det(R)

1.0

Why does this make sense? Discuss.

Is R invertible?  If so, what is its inverse?

In [12]:
inv(R)

array([[ 0.8660254,  0.5      ],
       [-0.5      ,  0.8660254]])

Use Worksheets A & B to determine what multiplying by the inverse of R does.  Why does this make sense?  Discuss.

Analyze S

Use Worksheet A to determine how multiplying by S changes areas.

Use Worksheet B to determine if one needs to flip out of the plane to go from the red F to the blue F.

Compute the determinant of S.

In [24]:
det(S)

1.0

Why does this make sense? Discuss.

Is S invertible?  If so, what is its inverse?

In [25]:
inv(S)

array([[ 1.,  0.],
       [-2.,  1.]])

Use Worksheets A & B to determine what multiplying by the inverse of S does.  Why does this make sense?  Discuss.

Analyze A

Use Worksheet A to approximately determine how multiplying by A changes areas.

Use Worksheet B to determine if one needs to flip out of the plane to go from the red F to the blue F.

Compute the determinant of A.

In [26]:
det(A)

-0.41526445641765725

Why does this make sense? Discuss.

Is A invertible?  If so, what is its inverse?

In [27]:
inv(A)

array([[-0.23357184,  1.30389852],
       [ 2.14416482, -1.65974349]])

Use Worksheets A & B to determine what multiplying by the inverse of S does.  Why does this make sense?  Discuss.

Analyze R*D1*R^(-1)

Define M to be R*D1*R^(-1)

In [28]:
M=R@D1@inv(R)

Use Worksheet A to determine how multiplying by M changes areas.  Why does this make sense?  Discuss.

Use Worksheet B to determine if one needs to flip out of the plane to go from the red F to the blue F.

Compute the determinant of M.

In [29]:
det(M)

1.4999999999999998

Why does this make sense? Discuss.

Is M invertible?  If so, what is its inverse?

In [30]:
inv(M)

array([[1.58333333, 0.72168784],
       [0.72168784, 0.75      ]])

Use Worksheets A & B to determine what multiplying by the inverse of M does.  Why does this make sense?  Discuss.

Compare M^(-1) to R*D1^(-1)*R^(-1).

In [31]:
np.allclose(inv(M),R@inv(D1)@inv(R))

True

Recall properties of matrix inverses that explain the result above. Discuss.

Let r1 and r2 be the first and second columns of R.

In [32]:
r1=R[:,0]
r2=R[:,1]

Multiply M(r1)

In [33]:
M@r1

array([0.4330127, 0.25     ])

We will compare M(r1) to r1 by dividing COMPONENT-WISE.  This is not a linear algebra operation from class, but it will help us see the relationship between the two vectors. 

In [34]:
M@r1/r1

array([0.5, 0.5])

and M(r2):

In [35]:
M@r2/r2

array([3., 3.])

Discuss what you see.  Why does this make sense?

Analyze a larger matrix.

In [36]:
G=np.diag(np.array([2, -1, 4, 0.5, 1]))
det(G)

-3.999999999999999

Why does the determinant make sense geometrically?  Discuss.

Exercises

1a. Define B=R*D3*R^(-1).

In [37]:
B=R@D3@inv(R)
print(B)

[[ 0.5        0.8660254]
 [ 0.8660254 -0.5      ]]


b. Using Worksheets A & B, determine how multiplying by B affects area and whether one needs to flip the red F out of the plane to get to the blue F.  Write what you find here.

Multiplying by B does not affect area, but it does reflect the shape so that one would need to move outside the plane to get to the blue F. It also seems to rotate the shape slightly.

c. Compute the determinant of B and explain why that agrees with your answer to 1b.


In [15]:
det(B)

-1.0

This agrees with my answer because the absolute value of the determinant is 1.0, which means the area is scaled by 1 (unchanged), and the determinant is negative, which means you would need to move outside the plane to complete the transformation.

d. Compute B(r1) and B(r2) and then explain the output.


In [57]:
print(B@r1)
print(B@r2)


[0.8660254 0.5      ]
[ 0.5       -0.8660254]


r1 is vector 30 degrees above x-axis. Multiplying it by B performs the operations from by from right to left. First, R^(-1) rotates r1 down 30 degrees, then D3 flips it over the x axis. Since it's already on the x-axis, this operation does nothing. Then, R rotates it back up 30 degrees, so it is back where it started (stays the same).

r2 is 30 degrees to the left the y-axis. It gets rotated 30 degrees clockwise, then flipped over the x-axis, then rotated 30 degrees counter-clockwise. This leaves it 30 degrees to the left of the bottom half of the y-axis.

2a. Define C=R*D4*R^(-1).


In [40]:
C=R@D4@inv(R)
print(C)

[[ 0.25      -0.4330127]
 [-0.4330127  0.75     ]]


b. Using Worksheets A & B, determine how multiplying by C affects area and whether one needs to flip the red F out of the plane to get to the blue F.  Write what you find here.


C makes the area 0 because the determinant is 0. The shape is flattened onto a line so moving inside/outside of the plane no longer matters.

c. Compute the determinant of C and explain why that agrees with your answer to 2b.


In [39]:
det(C)

-8.299974754062957e-18

This output agrees with my answer to 2b because the determinant is extremely close to zero (likely due to floating point arithmetic).

d. Compute C(r1) and C(r2) and then explain the output.


In [52]:
print(C@r1)
print(C@r2)

[-3.14741177e-17  6.93889390e-17]
[-0.5        0.8660254]


Similar to before, r1 gets rotated down 30 degrees by R^(-1), but D4 casts it onto the y-axis. Since before casting, it was aligned with the x-axis, it basically becomes a point at the origin. Since it is a point, rotating it up 30 degrees (R) does nothing, so the output is (0,0) at the origin (slightly off due to floating point error).

r2 gets rotated right 30 degrees so it is on the y-axis. Casting it onto itself does nothing, so D4 does nothing. Then, multiplying by R moves it back over 30 degrees to where it started, so it ends operations in the same position (unchanged).

3a. Define E=R*S*D3*D1.


In [51]:
E=R@S@D3@D1
print(E)

[[-0.0669873   1.5       ]
 [ 1.1160254  -2.59807621]]


b. Compute the determinant of E.


In [44]:
det(E)

-1.4999999999999998

c. Using Worksheets A & B, determine how multiplying by C affects area and whether one needs to flip the red F out of the plane to get to the blue F.  Explain how the answer to 3b looks reasonable given what you see.


Based on worksheet A, the area of the cube seems pretty similar, so a magnitude of 1.5 seems reasonable for the determinant. Using worksheet B, you can clearly see that the F has been reflected, so the negative determinant makes sense as well.